src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,1416 @@
+            {
+                return false;
+            }
+
+            //
+            // Check whether "Start" is a valid first occurrence
+            DateTimeOffset start = settings.Start.Value;
+
+            DateTime alignedStart = start.DateTime + GetRecurrenceTimeZone(settings) - start.Offset;
+
+            if (!pattern.DaysOfWeek.Any(day =>
+                    NthDayOfWeekInTheMonth(alignedStart, pattern.Index, day) == alignedStart.Date))
+            {
+                paramName = nameof(settings.Start);
+
+                reason = NotMatched;
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool TryValidateAbsoluteYearlyRecurrencePattern(TimeWindowFilterSettings settings, out string paramName, out string reason)
+        {
+            paramName = null;
+
+            reason = null;
+
+            RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+            TimeSpan intervalDuration = TimeSpan.FromDays(pattern.Interval * MinYearDayNumber);
+
+            //
+            // Time window duration must be shorter than how frequently it occurs
+            if (settings.End.Value - settings.Start.Value > intervalDuration)
+            {
+                paramName = $"{nameof(settings.End)}";
+
+                reason = OutOfRange;
+
+                return false;
+            }
+
+            //
+            // Required parameters
+            if (!TryValidateMonth(settings, out paramName, out reason))
+            {
+                return false;
+            }
+
+            if (!TryValidateDayOfMonth(settings, out paramName, out reason))
+            {
+                return false;
+            }
+
+            //
+            // Check whether "Start" is a valid first occurrence
+            DateTimeOffset start = settings.Start.Value;
+
+            DateTime alignedStart = start.DateTime + GetRecurrenceTimeZone(settings) - start.Offset;
+
+            if (alignedStart.Day != pattern.DayOfMonth.Value || alignedStart.Month != pattern.Month.Value)
+            {
+                paramName = nameof(settings.Start);
+
+                reason = NotMatched;
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool TryValidateRelativeYearlyRecurrencePattern(TimeWindowFilterSettings settings, out string paramName, out string reason)
+        {
+            paramName = null;
+
+            reason = null;
+
+            RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+            TimeSpan intervalDuration = TimeSpan.FromDays(pattern.Interval * MinYearDayNumber);
+
+            //
+            // Time window duration must be shorter than how frequently it occurs
+            if (settings.End.Value - settings.Start.Value > intervalDuration)
+            {
+                paramName = $"{nameof(settings.End)}";
+
+                reason = OutOfRange;
```

Comment:
Do we need more clear error info? like "Time window duration must be shorter than how frequently it occurs"

Reply:
Added.